Add gpg support to in-toto-record cli tool and library functions

lukpueh opened the pull request:
**Description of the changes being introduced by the pull request**:
This PR adds gpg signing capabilities to the `in-toto-record` command line tool and its corresponding library functions.
*Note: in-toto-record can be used to create link metadata for multipart steps*

Furthermore this PR makes the following enhancements in in-toto-record:
- Declutter the command line interface code (77cac9e5b9c6346ef49fc370d51840a97a2c6b18)
- Make in-toto-record's subcommands `start` and `stop` the first arguments. Shared and arguments specific to the subcommands come afterwards. Former is realized with a shared [`argparse` parent parser](https://docs.python.org/2.7/library/argparse.html#parents) (fa227d6e6b772f9209bb4ca5c6147893366c341d)
- Remove custom cli tool help message and improve auto help message with [`argparse` metavars](https://docs.python.org/2.7/library/argparse.html#metavar) (20ddbfdc7f8fc51521039a2abb40b048c86c9f28)
- Update README (7ad6d551261b1a1f822573e831e9f2d400e69c67)
- Refactor unit tests (5ced4b8bac49372ed07f64012a54c9c2e71f7abd)

**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

Commit:
Refactor/remove functions in in-toto-record

This commit removes functions in_toto_record_start and
in_toto_record_stop from the command line tool in-toto-record
since they just used to pass on arguments to the corresponding
functions in runlib but don't add any functionality. Keeping them
around just makes changes to the API more cumbersome.

Commit:
Refactor runlib.in_toto_record* arg order

This commit also refactors `runlib.in_toto_record_start` and
`runlib.in_toto_record_start` parameters. The `key` argument is
now the last argument and is called `signing_key`.
I do this because I will add additional key related arguments and
it seems neater to list them next to each other.

Commit:
Add parent parser to in-toto-record cli tool

The parent parser is used to inherit shared arguments to both
subparsers, it also changes the order of arguments passed to the
cli tool. The subcommands are now the first arguments, i.e.
in-toto-record {start, stop} [other arguments]

Commit:
Remove custom help message from in-toto-record

This commit adds metavar argument descriptions to make the
default help text more meaningful. As a consequence we can remove
the custom help text.

Commit:
Add gpg support to in-toto-record cli tool and lib

Note that for runlib.in_toto_record_stop we have to apply a
workaround if we want to use the default gpg key to be able
to load and verify the preliminary link file created with
in_toto_record_start, because the keyid required to load and
verify the right file is not available beforehand.
As a workaround we glob for any preliminary files (and fail if we
don't find exactly one) and then use the keyid from the files
signature.

Commit:
Update in-toto-record README section and doc header

Adds information about gpg signing and generally rephrases the
in-toto-record section in README.md and in_toto-record.py's doc
header.

Commit:
Refactor in-toto-record tests and add exit 0

Refactors unit tests for in_toto_record.py and adds a sys.exit(0)
if the command line tool runs successfully, which is convenient
for testing.

Commit:
Add in-toto-record gpg tests

Commit:
Add runlib.in_toto_record* key arg tests

Commit:
Fix test in gpg record stop

Fix an in-toto-record stop test that is only executed if the
available gpg version is not fully supported.

The test should not fail, but it requires a preliminary link file
which in turn requires a call to in-toto-record start, passing
a gpg keyid.

See code comments for more info.

SantiagoTorres commented on in_toto/in_toto_record.py:
```diff
@@ -14,166 +13,141 @@
+
+  # Otherwise we interpret it as actual keyid
+  gpg_keyid = None
+  if args.gpg != True:
```
should this be
```python
if not gpg_use_default and args.gpg:
  ...
```
or so?

lukpueh replied:
Works either way, but yours is definitely more readable. I'll update and write a better comment.

SantiagoTorres commented on in_toto/in_toto_record.py:
```diff
@@ -14,166 +13,141 @@
 <Purpose>
-  Provides a command line interface to start and stop in-toto link metadata
-  recording.
+  Provides a command line interface to create a link metadata file in multiple
```
What do you think of changing "multiple" to "two"?

lukpueh replied:
Hm. I was actually referring to the (potentially) multiple steps in between `start` and `record,`. But you are right, this is not clear. What about something like:
*"...create a link metadata file in two steps, in order to provide evidence for arbitrary activities. Note: If you want to provide evidence for a single command take a look at in-toto-record."*?

SantiagoTorres commented on in_toto/in_toto_record.py:
```diff
@@ -14,166 +13,141 @@
+
   # We load the key here because it might prompt the user for a password in
   # case the key is encrypted. Something that should not happen in the library.
+  key = None
```
AFAIR, we need a ward that key is not used along gpg right? Is this being done by argparse or do we need to add it.

lukpueh replied:
Right, I'm using an argparse mutually exclusive group for `key` and `gpg` above. So this is taken care of.

SantiagoTorres commented on in_toto/in_toto_record.py:
```diff
@@ -14,166 +13,141 @@
+    try:
+      key = in_toto.util.prompt_import_rsa_key_from_file(args.key)
+
+    except Exception as e:
```
👍 for maximizing code reuse. I think we can add specific exception handling code here though. What do you think?

lukpueh replied:
Yes. This applies to all command line tools. I made a note to fix this in a separate PR https://github.com/in-toto/in-toto/issues/126#issuecomment-356707753

SantiagoTorres commented on in_toto/in_toto_record.py:
```diff
@@ -14,166 +13,141 @@
+          signing_key=key, gpg_keyid=gpg_keyid,
+          gpg_use_default=gpg_use_default, gpg_home=args.gpg_home)
+
+    elif args.command == "stop": # pragma: no branch
```
I could go both ways, but if argparse already catches args.command == "start" || "stop" we could just use an else here no?

lukpueh replied:
I had the same thoughts (in various places throughout the repo). I am not sure about *the right way* to do this. For the sake of readability something like `else: # args.command == "stop"` might be a good compromise. :)

lukpueh replied:
FYI https://github.com/secure-systems-lab/code-style-guidelines/issues/18

SantiagoTorres commented on in_toto/runlib.py:
```diff
@@ -487,34 +522,70 @@ def in_toto_record_start(step_name, key, material_list):
+    log.info("Signing link metadata using passed key...")
+    signature = link_metadata.sign(signing_key)
+
+  elif gpg_keyid:
```
I think we can take the `else` away by having `elif gpg_keyid or gpg_use_default:`

lukpueh replied:
Or even `else: # gpg_keyid or gpg_use_default`, c.f. https://github.com/in-toto/in-toto/pull/167#discussion_r160774560 :)

lukpueh replied:
Oh, the reason why I did it that way was to have different log statements. "*using passed GPG keyid"* vs. "*using default GPG key"*.
I could put the function call which is the same for both in an `else`, and have an additional `if/else` inside for the log statement, but I don't think that that would increase readability.

vladimir-v-diaz replied:
Another option is to relocate the log statement to `sign_gpg()`.  It might be easier to handle the "GPG keyid" vs "default GPG key" case there.

Note: The guidelines recommend never to use short circuit evaluation to produce output.
https://github.com/secure-systems-lab/code-style-guidelines#never-use-short-circuit-evaluation-to-produce-output

SantiagoTorres replied:
Hmm, I think the line is blurry on what is "short circuit evaluation" then. This is just an `or` clause. Am I missing something?

SantiagoTorres replied:
re "custom message for default keyid". I don't know how useful is it really. We can leave it though...

lukpueh replied:
Alright, I added a pointer https://github.com/in-toto/in-toto/issues/117#issuecomment-356756728. Thanks for the review!

SantiagoTorres commented on test/test_in_toto_record.py:
```diff
@@ -43,101 +41,135 @@ class TestInTotoRecordTool(unittest.TestCase):
+      # Fail (exit 1) start recording on not fully supported gpg version
+      self._test_cli_sys_exit(["start"] + args, 1)
+
+      # NOTE: Stop recording with default gpg on not fully supported gpg
```
I don't think this comment is clear enough...

lukpueh replied:
Ok, I'll reword.

Commit:
Rephrase purpose doc header in in-toto-record

Commit:
Fix comments and readability in in_toto_record

Reword comments and update conditionals for better readability
addressing @SantiagoTorres's review comments.

SantiagoTorres commented:
LGTM, other than that open discussion regarding the short-circuit evaluation.
